### DISCUSSION TITLE
Ternary operators

### DISCUSSION THREAD

**Discussion by iamwizzdom** [smarty-php/smarty#660] [Ideas]
Please implement support for ternary operators such as $foo ?: null, $foo?->prop, $foo ? $var1 : $var2
These ternary operators are very helpful in simplifying code. Thanks in advance ☺️

**Comment by tpneumat** [comment]
bump - I want this too for templates.  Example with "elvis" operator

`{if $myvar}{$myvar}{else}N/A{/if}`

could be simplified to

`{$myvar?:'N/A'}`

Thx

**Comment by scottchiefbaker** [comment]
Can't this already be done with "default"?`{$myvar|default:"foobar"}`On Apr 30, 2023 7:05 AM, tpneumat ***@***.***> wrote:
bump - I want this too for templates.  Example with "elvis" operator
{if $myvar}{$myvar}{else}N/A{/if}
could be simplified to
{$myvar?:'N/A'}
Thx

**Reply by tpneumat** [reply]
I suppose so, still that's a longer syntax.  But what if you wanted to do this?
`{$myvar?'was set':'not set'}`

That has to be done with an "if, else, end if" far as I am aware.
`{if $myvar}was set{else}not set{/if}`

**Comment by wisskid** [comment]
Funny thing is that Smarty supports the ternary since 2009/2010. But it was never documented. Also, it requires the testing expression to be in parentheses. So this should work:
```smarty
{($foo) ? $var1 : $var2}
```
But this will not:
```smarty
{$foo ? $var1 : $var2}
```

There is no support for the shorthand `?:` version. Nor for the null coalescing operator `??`. We could try improving support (and docs) for the ternary and maybe also try supporting the null coalescing operator.

See: #881 and #882.

**Reply by Grawl** [reply]
can confirm it works

**Comment by wisskid** [comment]
Ternary support is implemented in v5

**Comment by tpneumat** [comment]
Thank you!
